sklearnex/covariance/incremental_covariance.py

```diff
@@ -293,21 +328,34 @@ def _onedal_fit(self, X, queue=None):
         return self
 
     # expose sklearnex pairwise_distances if mahalanobis distance eventually supported
-    @wrap_output_data
     def mahalanobis(self, X):
         if sklearn_check_version("1.0"):
-            self._validate_data(X, reset=False, copy=self.copy)
-        else:
-            check_array(X, copy=self.copy)
+            self._check_feature_names(X, reset=False)
 
+        xp, _ = get_namespace(X)
         precision = self.get_precision()
-        with config_context(assume_finite=True):
-            # compute mahalanobis distances
-            dist = pairwise_distances(
-                X, self.location_[np.newaxis, :], metric="mahalanobis", VI=precision
-            )
+        # compute mahalanobis distances
+        # pairwise_distances will check n_features (via n_feature matching with
+        # self.location_) , and will check for finiteness via check array
+        # check_feature_names will match _validate_data functionally
+        location = self.location_[np.newaxis, :]
+        if "numpy" not in xp.__name__:
+            location = xp.asarray(location, device=X.device)
+
+        try:
+            dist = pairwise_distances(X, location, metric="mahalanobis", VI=precision)
+        except ValueError as e:
+            # Throw the expected sklearn error in an n_feature length violation
+            if "Incompatible dimension for X and Y matrices: X.shape[1] ==" in str(e):
+                n_features = re.findall(r"\s\d+", str(e))[0].lstrip()
```

Comment:
Is this really necessary? :)

Reply:
Unfortunately yes, its catching a specific error and value from a raised error in pairwise_distances and throwing an error like _validate_data. This is necessary for sklearn test_conformance.

Reply:
Swapped to _num_features from sklearn instead. Good call.